ssawg_trending_scraper.py

```diff
@@ -0,0 +1,376 @@
+import netrc
+import argparse
+import Ska.ftp
+from bs4 import BeautifulSoup
+from urllib import request, response, error, parse
+from urllib.request import urlopen
+from datetime import datetime, timedelta
+import Chandra.Time
+import re
+
+
+# ------------------------------------
+# Setup for password-protected site(s)
+# ------------------------------------
+
+netrc = Ska.ftp.parse_netrc()
+if 'https://cxc.cfa.harvard.edu' not in netrc:
+    raise RuntimeError('must have authentication in .netrc')
+
+user = netrc['https://cxc.cfa.harvard.edu']['login']
+password = netrc['https://cxc.cfa.harvard.edu']['password']
+
+# -----------------------------------
+# ESTABLISH BASE URL, TRENDING PAGES
+# AND CREATE PROPER URLS TO USE
+# -----------------------------------
+
+base_url = "https://cxc.cfa.harvard.edu/mta/ASPECT/"
+
+trending_pages = [
+    "kalman_watch/", 
+    "acq_stat_reports/",
+    "gui_stat_reports/",
+    "perigee_health_plots/",
+    "periscope_drift_reports/", # this page requires a username/password
+    "obc_rate_noise/trending/",
+    "fid_drift/",
+    "aimpoint_mon/",
+    "celmon/",
+    "vv_rms/",
+    "attitude_error_mon/",
+    "fss_check3/"
+]
+
+
+def proper_url(page, base_url):
+    if page[-8:-1] == "reports": # conditional to perfrom unique checks on acq/gui stat reports
+        for offset in range(4, 0, -1):
+            year = (datetime.now()).year
+            url = (f'{base_url}{page}{year}/Q{offset}/index.html') # creates the temporary url; will later need '/index.html' removed
+            if page == "periscope_drift_reports/": # this page requires a username/password
+                page_request = requests.get(url, auth=(user, password))
+            else:
+                page_request = requests.get(url)
+            if page_request.status_code < 300: # verify successful page
+                soup = BeautifulSoup(page_request.text, "lxml")  
+                temp_tables = soup.find_all('table') # locate table with quarterly dates
+                start_time = temp_tables[1].text[12:34] # quarterly start date
+                stop_time = temp_tables[1].text[34:] # quarterly stop date
+                stop_minus_start = Chandra.Time.date2secs(stop_time) - Chandra.Time.date2secs(start_time)
+                halfway = Chandra.Time.date2secs(start_time) + stop_minus_start
+                if Chandra.Time.date2secs(datetime.now()) > halfway: # if 50% through quarter
+                    return (f'{base_url}{page}{year}/Q{offset}/')
+                elif offset == 1: # if it's the first quarter of the year
+                    return (f'{base_url}{page}{year-1}/Q4/') # switch to fourth quarter of previous year
+                else:
+                    return (f'{base_url}{page}{year}/Q{offset-1}/') # try previous quarter
+            else:
+                continue
+    elif page ==  "perigee_health_plots/": # perform unique checks on this page
+        for offset in range(0, -31, -1): # looks back one month
```

Comment:
This block (and the if block above) should be factored out into separate functions so the flow is clear.  This basically turns this function into a dispatcher, which makes one think that you might end up doing this with subclasses.
```
if <it is a reports page>:
    out = get_proper_url_reports(..)
elif <it is a perigee page>:
    out = get_proper_url_perigee(..)
else:
    out = get_proper_url_generic(..)
return out
```

Reply:
Yeah, as it grew, I knew I should probably separate them but wasn't sure which was better. I'll pop them out into their own functions.